docs/specification/draft/client/elicitation.mdx

```diff
@@ -321,12 +526,212 @@ Servers should handle each state appropriately:
+    User->>Client: Provide consent
+    Client->>UserAgent: Open URL
+    Client->>Server: Accept response
+    Client-->>Server: elicitation/track request (optional)
+    UserAgent->>3AS: Load authorize route
+    Note over 3AS,User: User interaction (OAuth flow):<br>User consents to scoped MCP Server access
+    3AS->>UserAgent: redirect to MCP Server's redirect_uri
+    UserAgent->>Server: load redirect_uri page
+    Note over Server: Confirm: redirect_uri belongs to MCP Server
+    Note over Server: Confirm: elicitation matches user session<br>Confirm: user is logged into MCP Server or MCP AS
+    Server->>3AS: Exchange authorization code for  OAuth tokens
+    3AS->>Server: Grants tokens
+    Note over Server: Bind tokens to MCP user identity
+    Server-->>Client: elicitation/track response (optional)
+    Client->>Server: Retry tools/call
+    Note over Server: Retrieve token bound to user identity
+    Server->>3RS: Call 3rd-party API
+```
+
+This pattern maintains clear security boundaries while enabling rich integrations with third-party services that require user authorization.
+
+## Error Handling
+
+Servers **SHOULD** return standard JSON-RPC errors for common failure cases:
+
+- Unknown `elicitationId` in an `elicitation/track` request: `-32602` (Invalid params)
+- An `elicitationId` belonging to a different client in an `elicitation/track` request: `-32602` (Invalid params)
+- When a request cannot be processed until an elicitation is completed: `-32604` (ElicitationRequired)
+
+Clients **SHOULD** return standard JSON-RPC errors for common failure cases:
+
+- Server sends an `elicitation/create` request with a mode not declared in client capabilities: `-32602` (Invalid params)
+
 ## Security Considerations
 
-1. Servers **MUST NOT** request sensitive information through elicitation
+1. Clients **MUST** provide clear indication of which server is requesting information
 2. Clients **SHOULD** implement user approval controls
-3. Both parties **SHOULD** validate elicitation content against the provided schema
-4. Clients **SHOULD** provide clear indication of which server is requesting information
-5. Clients **SHOULD** allow users to decline elicitation requests at any time
-6. Clients **SHOULD** implement rate limiting
-7. Clients **SHOULD** present elicitation requests in a way that makes it clear what information is being requested and why
+3. Clients **SHOULD** allow users to decline elicitation requests at any time
+4. Clients **SHOULD** implement rate limiting
+5. Clients **SHOULD** present elicitation requests in a way that makes it clear what information is being requested and why
+6. Servers **MUST** bind elicitation requests to the user's identity
+
+### Safe URL Handling
+
+Clients implementing url elicitation **MUST** handle URLs carefully to prevent users from unknowingly clicking malicious links.
+
+1. Servers **MUST NOT** include URLs in any message or schema fields as part of a form elicitation request.
```

Comment:
Why not? This feels like an odd restriction. Does it mean I can't include a URL in an elicitation message?

Reply:
> Does it mean I can't include a URL in an elicitation message?

Correct, that's my intention here.

My thinking is: By saying that URLs _must_ only be present in the `url` field (not in `message`), clients can aggressively enforce it. A rogue server can of course send whatever message it wants, but a spec-compliant client can and should choose not to make the an extra URL in the message click-able, to further force the user to properly and knowingly consent to only clicking the elicited URL.

If you think that is too much of a stretch, I'm happy to adjust it.